File: MSHFConfig.m
#import "public/MSHFConfig.h"
#import "public/MSHFUtils.h"
#import <libcolorpicker.h>

@interface DarwinNotificationsManager : NSObject

@property (strong, nonatomic) id someProperty;

+ (instancetype)sharedInstance;

- (void)registerForNotificationName:(NSString *)name callback:(void (^)(void))callback;
- (void)postNotificationWithName:(NSString *)name;

@end

@implementation MSHFConfig

- (instancetype)initWithDictionary:(NSDictionary *)dict {
  self = [super init];
  if (self) {
    [self setDictionary:dict];
    [[DarwinNotificationsManager sharedInstance] registerForNotificationName:MSHFPreferencesChanged callback:^() {
      [self reloadConfig];
    }];
  }
  return self;
}

- (void)initializeViewWithFrame:(CGRect)frame {
  UIView *superview = nil;
  NSUInteger index;

  if (_view) {
    if ([_view superview]) {
      superview = [_view superview];
      index = [superview.subviews indexOfObject:_view];
    }

    [_view stop];
    [_view removeFromSuperview];
  }

  switch (_style) {
  case 1:
    self.view = [[MSHFBarView alloc] initWithFrame:frame];
    [((MSHFBarView *)_view) setBarSpacing:self.barSpacing];
    [((MSHFBarView *)_view) setBarCornerRadius:self.barCornerRadius];
    break;
  case 2:
    self.view = [[MSHFLineView alloc] initWithFrame:frame];
    [((MSHFLineView *)_view) setLineThickness:self.lineThickness];
    break;
  case 3:
    self.view = [[MSHFDotView alloc] initWithFrame:frame];
    [((MSHFDotView *)_view) setBarSpacing:self.barSpacing];
    break;
  case 4:
    self.view = [[MSHFSiriView alloc] initWithFrame:frame];
    break;
  default:
    self.view = [[MSHFJelloView alloc] initWithFrame:frame];
  }

  if (superview) {
    if (index == NSNotFound) {
      [superview addSubview:_view];
    } else {
      [superview insertSubview:_view atIndex:index];
    }
  }

  [self configureView];
}

- (void)configureView {
  _view.autoHide = self.enableAutoHide;
  _view.displayLink.preferredFramesPerSecond = self.fps;
  _view.numberOfPoints = self.numberOfPoints;
  _view.waveOffset = self.waveOffset + self.waveOffsetOffset;
  _view.gain = self.gain;
  _view.limiter = self.limiter;
  _view.sensitivity = self.sensitivity;
  _view.audioProcessing.fft = self.enableFFT;
  _view.disableBatterySaver = self.disableBatterySaver;
  NSLog(@"[Mitsuha] self.waveColor: %@", self.waveColor);
  NSLog(@"[Mitsuha] self.subwaveColor: %@", self.subwaveColor);
  NSLog(@"[Mitsuha] self.subSubwaveColor: %@", self.subSubwaveColor);
  NSLog(@"[Mitsuha] self.colorMode: %d", self.colorMode);
  
  _view.siriEnabled = self.colorMode == 1;

  if (self.colorMode == 2 && self.waveColor) {
    if (self.style == 4) {
      [_view updateWaveColor:[self.waveColor copy]
                subwaveColor:[self.waveColor copy]
             subSubwaveColor:[self.waveColor copy]];
    } else {
      [_view updateWaveColor:[self.waveColor copy]
                subwaveColor:[self.waveColor copy]];
    }
  } else if (self.colorMode == 1 && self.waveColor && self.subwaveColor && self.subSubwaveColor) {
    [_view updateWaveColor:[self.waveColor copy]
              subwaveColor:[self.subwaveColor copy]
           subSubwaveColor:[self.subSubwaveColor copy]];
  } else if (self.calculatedColor) {
    [_view updateWaveColor:[self.calculatedColor copy]
              subwaveColor:[self.calculatedColor copy]];
  }
}
- (UIColor *)getAverageColorFrom:(UIImage *)image withAlpha:(double)alpha {
  CGSize size = {1, 1};
  UIGraphicsBeginImageContext(size);
  CGContextRef ctx = UIGraphicsGetCurrentContext();
  CGContextSetInterpolationQuality(ctx, kCGInterpolationMedium);

  [image drawInRect:(CGRect){.size = size} blendMode:kCGBlendModeCopy alpha:1];

  uint8_t *data = (uint8_t *)CGBitmapContextGetData(ctx);

  UIColor *color = [UIColor colorWithRed:data[2] / 255.0f
                                   green:data[1] / 255.0f
                                    blue:data[0] / 255.0f
                                   alpha:alpha];

  UIGraphicsEndImageContext();
  return color;
}
- (void)colorizeView:(UIImage *)image {
  if (self.view == NULL)
    return;
  UIColor *color = self.waveColor;
  if (self.colorMode != 2) {
    
    UIColor *scolor = self.waveColor;
    UIColor *sscolor = self.waveColor;
    
    if (self.colorMode == 1) {
      color = [UIColor colorWithRed:1.0f
                              green:0.0f
                              blue:0.0f
                              alpha:self.dynamicColorAlpha];
      scolor = [UIColor colorWithRed:0.0f
                              green:1.0f
                                blue:0.0f
                              alpha:self.dynamicColorAlpha];
      sscolor = [UIColor colorWithRed:0.0f
                                green:0.0f
                                blue:1.0f
                                alpha:self.dynamicColorAlpha];
    } else {
      color = [self getAverageColorFrom:image
                              withAlpha:self.dynamicColorAlpha];
      self.calculatedColor = color;
    }
    
    if (self.colorMode == 1) {
      [self.view updateWaveColor:[color copy]
                    subwaveColor:[scolor copy]
                subSubwaveColor:[sscolor copy]];
    } else if (self.colorMode != 1 && self.style == 4) {
      [self.view updateWaveColor:[color copy]
                    subwaveColor:[color copy]
                subSubwaveColor:[color copy]];
    }  else {
      [self.view updateWaveColor:[color copy]
                    subwaveColor:[color copy]];
    }
  } else {
    if (self.style == 4) {
      [_view updateWaveColor:[color copy]
                subwaveColor:[color copy]
             subSubwaveColor:[color copy]];
    } else {
      [_view updateWaveColor:[color copy]
                subwaveColor:[color copy]];
    }
  }
}

- (void)setDictionary:(NSDictionary *)dict {
  _application = [dict objectForKey:@"application"];
  _enabled = [([dict objectForKey:@"enabled"] ?: @(YES)) boolValue];

  _enableDynamicGain =
      [([dict objectForKey:@"enableDynamicGain"] ?: @(NO)) boolValue];
  _style = [([dict objectForKey:@"style"] ?: @(0)) intValue];
  _colorMode = [([dict objectForKey:@"colorMode"] ?: @(0)) intValue];
  _enableAutoUIColor =
      [([dict objectForKey:@"enableAutoUIColor"] ?: @(YES)) boolValue];
  _ignoreColorFlow =
      [([dict objectForKey:@"ignoreColorFlow"] ?: @(NO)) boolValue];
  _enableCircleArtwork =
      [([dict objectForKey:@"enableCircleArtwork"] ?: @(NO)) boolValue];
  _enableCoverArtBugFix =
      [([dict objectForKey:@"enableCoverArtBugFix"] ?: @(NO)) boolValue];
  _disableBatterySaver =
      [([dict objectForKey:@"disableBatterySaver"] ?: @(NO)) boolValue];
  _enableFFT = [([dict objectForKey:@"enableFFT"] ?: @(NO)) boolValue];
  _enableAutoHide =
      [([dict objectForKey:@"enableAutoHide"] ?: @(YES)) boolValue];

  if ([dict objectForKey:@"waveColor"]) {
    if ([[dict objectForKey:@"waveColor"] isKindOfClass:[UIColor class]]) {
      _waveColor = [dict objectForKey:@"waveColor"];
    } else if ([[dict objectForKey:@"waveColor"]
                   isKindOfClass:[NSString class]]) {
      _waveColor =
          LCPParseColorString([dict objectForKey:@"waveColor"], @"#000000:0.5");
    } else {
      _waveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    }
  } else {
    _waveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
  }

  if ([dict objectForKey:@"subwaveColor"]) {
    if ([[dict objectForKey:@"subwaveColor"] isKindOfClass:[UIColor class]]) {
      _subwaveColor = [dict objectForKey:@"subwaveColor"];
    } else if ([[dict objectForKey:@"subwaveColor"]
                   isKindOfClass:[NSString class]]) {
      _subwaveColor = LCPParseColorString([dict objectForKey:@"subwaveColor"],
                                          @"#000000:0.5");
    } else {
      _subwaveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    }
  } else {
    _subwaveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
  }
  
  if ([dict objectForKey:@"subSubwaveColor"]) {
    if ([[dict objectForKey:@"subSubwaveColor"] isKindOfClass:[UIColor class]]) {
      _subwaveColor = [dict objectForKey:@"subSubwaveColor"];
    } else if ([[dict objectForKey:@"subSubwaveColor"]
                   isKindOfClass:[NSString class]]) {
      _subwaveColor = LCPParseColorString([dict objectForKey:@"subSubwaveColor"],
                                          @"#000000:0.5");
    } else {
      _subwaveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    }
  } else {
    _subwaveColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
  }

  _gain = [([dict objectForKey:@"gain"] ?: @(50)) doubleValue];
  _limiter = [([dict objectForKey:@"limiter"] ?: @(0)) doubleValue];
  _numberOfPoints =
      [([dict objectForKey:@"numberOfPoints"] ?: @(8)) unsignedIntegerValue];
  _sensitivity = [([dict objectForKey:@"sensitivity"] ?: @(1)) doubleValue];
  _dynamicColorAlpha =
      [([dict objectForKey:@"dynamicColorAlpha"] ?: @(0.6)) doubleValue];

  _barSpacing = [([dict objectForKey:@"barSpacing"] ?: @(5)) doubleValue];
  _barCornerRadius =
      [([dict objectForKey:@"barCornerRadius"] ?: @(0)) doubleValue];
  _lineThickness = [([dict objectForKey:@"lineThickness"] ?: @(5)) doubleValue];

  _waveOffset = [([dict objectForKey:@"waveOffset"] ?: @(0)) doubleValue];
  _waveOffset = ([([dict objectForKey:@"negateOffset"] ?: @(false)) boolValue]
                     ? _waveOffset * -1
                     : _waveOffset);

  _fps = [([dict objectForKey:@"fps"] ?: @(24)) doubleValue];
}

+ (NSDictionary *)parseConfigForApplication:(NSString *)name {
  NSMutableDictionary *prefs = [@{} mutableCopy];
  [prefs setValue:name forKey:@"application"];

  NSDictionary *file =
      [[NSDictionary alloc] initWithContentsOfFile:MSHFPrefsFile];

  for (NSString *key in [file allKeys]) {
    [prefs setValue:[file objectForKey:key] forKey:key];
  }

  for (NSString *key in [prefs allKeys]) {
    NSString *removedKey = [key stringByReplacingOccurrencesOfString:[NSString stringWithFormat:@"MSHF%@", name] withString:@""];
    NSString *lowerCaseKey = [NSString stringWithFormat:@"%@%@",[[removedKey substringToIndex:1] lowercaseString], [removedKey substringFromIndex:1]];

    [prefs setValue:[prefs objectForKey:key] forKey:lowerCaseKey];
  }

  prefs[@"gain"] = [prefs objectForKey:@"gain"] ?: @(50);
  prefs[@"subwaveColor"] = prefs[@"waveColor"];
  prefs[@"subSubwaveColor"] = prefs[@"waveColor"];
  prefs[@"waveOffset"] = ([prefs objectForKey:@"waveOffset"] ?: @(0));

  return prefs;
}

- (void)reloadConfig {
  int oldStyle = self.style;
  [self setDictionary:[MSHFConfig parseConfigForApplication:self.application]];
  if (self.view) {
    if (self.style != oldStyle) {
      [self initializeViewWithFrame:self.view.frame];
      [[self view] start];
    } else {
      [self configureView];
    }
  }
}

+ (MSHFConfig *)loadConfigForApplication:(NSString *)name {
  return [[MSHFConfig alloc]
      initWithDictionary:[MSHFConfig parseConfigForApplication:name]];
}

@end

@implementation DarwinNotificationsManager {
    NSMutableDictionary * handlers;
}

+ (instancetype)sharedInstance {
    static id instance = NULL;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        handlers = [NSMutableDictionary dictionary];
    }
    return self;
}

- (void)registerForNotificationName:(NSString *)name callback:(void (^)(void))callback {
    handlers[name] = callback;
    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterAddObserver(center, (__bridge const void *)(self), defaultNotificationCallback, (__bridge CFStringRef)name, NULL, CFNotificationSuspensionBehaviorDeliverImmediately);
}

- (void)postNotificationWithName:(NSString *)name {
    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterPostNotification(center, (__bridge CFStringRef)name, NULL, NULL, YES);
}

- (void)notificationCallbackReceivedWithName:(NSString *)name {
    void (^callback)(void) = handlers[name];
    callback();
}

void defaultNotificationCallback (CFNotificationCenterRef center,
                 void *observer,
                 CFStringRef name,
                 const void *object,
                 CFDictionaryRef userInfo)
{
    NSLog(@"name: %@", name);
    NSLog(@"userinfo: %@", userInfo);

    NSString *identifier = (__bridge NSString *)name;
    [[DarwinNotificationsManager sharedInstance] notificationCallbackReceivedWithName:identifier];
}


- (void)dealloc {
    CFNotificationCenterRef center = CFNotificationCenterGetDarwinNotifyCenter();
    CFNotificationCenterRemoveEveryObserver(center, (__bridge const void *)(self));
}


@end

File: MSHFConfig.swift
import UIKit

private extension UIColor {
    class func hbcp_propertyList(value: String?) -> UIColor? {
        if var string = value {
            if string.count == 4 || string.count == 5 {
                let r = String(repeating: string[string.index(string.startIndex, offsetBy: 1)], count: 2)
                let g = String(repeating: string[string.index(string.startIndex, offsetBy: 2)], count: 2)
                let b = String(repeating: string[string.index(string.startIndex, offsetBy: 3)], count: 2)
                let a = string.count == 5 ? String(repeating: string[string.index(string.startIndex, offsetBy: 4)], count: 2) : "FF"
                string = String(format: "%@%@%@%@", r, g, b, a)
            }

            var hex: UInt64 = 0
            let scanner = Scanner(string: string)
            scanner.charactersToBeSkipped = CharacterSet(charactersIn: "#")
            scanner.scanHexInt64(&hex)
            if string.count == 9 {
                return self.init(
                    red: CGFloat((hex & 0xFF000000) >> 24) / 255,
                    green: CGFloat((hex & 0x00FF0000) >> 16) / 255,
                    blue: CGFloat((hex & 0x0000FF00) >> 8) / 255,
                    alpha: CGFloat((hex & 0x000000FF) >> 0) / 255)
            } else {
                var alpha: Float = 1
                if scanner.scanString(":", into: nil) {
					scanner.scanFloat(&alpha)
                }

                return self.init(
                    red: CGFloat((hex & 0xFF0000) >> 16) / 255,
                    green: CGFloat((hex & 0x00FF00) >> 8) / 255,
                    blue: CGFloat((hex & 0x0000FF) >> 0) / 255,
                    alpha: CGFloat(alpha))
            }
        }
        return nil
    }
}
        
private func LCPParseColorString(_ hexString: String?, _ fallback: String) -> UIColor {
    if let result = UIColor.hbcp_propertyList(value:hexString) {
        return result
    }
    else {
        return UIColor.hbcp_propertyList(value:fallback) ?? UIColor.black
    }
}


@objc (MSHFConfig) final public class MSHFConfig: NSObject {

    @objc private var enabled = false
    private var application: String?
    @objc private var style = 0
    @objc private var colorMode = 0
    private var enableAutoUIColor = false
    @objc private var enableCoverArtBugFix = false
    private var disableBatterySaver = false
    private var enableFFT = false
    private var enableAutoHide = false
    private var gain: Double = 50.0
    private var limiter: Double = 0.0
    private var waveColor: UIColor?
    private var subwaveColor: UIColor?
    private var subSubwaveColor: UIColor?
    private var calculatedColor: UIColor?
    private var numberOfPoints: Int = 0
    private var fps: Int = 24
    @objc private var waveOffset: CGFloat = 0.0
    @objc private var waveOffsetOffset: CGFloat = 0.0
    private var sensitivity: CGFloat = 0.0
    private var dynamicColorAlpha: CGFloat = 0.0
    private var barSpacing: CGFloat = 0.0
    private var barCornerRadius: CGFloat = 0.0
    private var lineThickness: CGFloat = 0.0
    @objc private var ignoreColorFlow = false
    @objc private var view: MSHFView?
    
    init(dictionary dict: Dictionary<String, Any>) {
        super.init()
        setDictionary(dict)
   
        // let MSHFPreferencesChanged =  "com.ryannair05.mitsuhaforever/ReloadPrefs"
    }
    
    @objc public func initializeView(withFrame frame: CGRect) {
        var superview: UIView?
        var index: Int?

        if let view = view {
            superview = view.superview
            index = superview?.subviews.firstIndex(of: view)

            view.stop()
            view.removeFromSuperview()
        }

        switch style {
            case 1:
                view = MSHFBarView(frame: frame)
                (view as! MSHFBarView).barSpacing = barSpacing
                (view as! MSHFBarView).barCornerRadius = barCornerRadius
            case 2:
                view = MSHFLineView(frame: frame)
                (view as! MSHFLineView).lineThickness = lineThickness
            case 3:
                view = MSHFDotView(frame: frame)
                (view as! MSHFDotView).barSpacing = barSpacing
            case 4:
                view = MSHFSiriView(frame: frame)
            default:
                view = MSHFJelloView(frame: frame)
        }

        if let superview = superview {
            if let index = index  {
                superview.insertSubview(view!, at: index)
            }
            else {
                superview.addSubview(view!)
            }
        }

        configureView()
    }

    private func configureView() {
        view!.autoHide = enableAutoHide
        view!.displayLink!.preferredFramesPerSecond = fps
        view!.numberOfPoints = numberOfPoints
        view!.waveOffset = waveOffset + waveOffsetOffset
        view!.gain = gain
        view!.limiter = limiter
        view!.sensitivity = sensitivity
        view!.audioProcessing?.fft = enableFFT
        view!.disableBatterySaver = disableBatterySaver
        view!.siriEnabled = colorMode == 1

        if let waveColor = waveColor {
            if colorMode == 2 {
                view!.updateWave(waveColor, subwaveColor: waveColor)
            } else if let subwaveColor = subwaveColor, let subSubwaveColor = subSubwaveColor, colorMode == 1 {
                view!.updateWave(waveColor,
                    subwaveColor: subwaveColor,
                    subSubwaveColor: subSubwaveColor)
            }
        }
        else if let calculatedColor = calculatedColor {
            view!.updateWave(calculatedColor, subwaveColor: calculatedColor)
        }
    }

    private func getAverageColor(from image: UIImage, withAlpha alpha: CGFloat) -> UIColor {
        let size = CGSize(width: 1, height: 1)
        UIGraphicsBeginImageContext(size)
        let ctx = UIGraphicsGetCurrentContext()
        ctx!.interpolationQuality = .medium

        image.draw(in: CGRect(x: 0, y: 0, width: 1, height: 1), blendMode: .copy, alpha: 1)

        let data = Array(UnsafeBufferPointer(start: ctx?.data?.bindMemory(to: UInt8.self, capacity: 4), count: 4)).map{CGFloat($0)}

        let color = UIColor(
            red: data[2] / 255.0,
            green: data[1] / 255.0,
            blue: data[0] / 255.0,
            alpha: alpha)

        UIGraphicsEndImageContext()
        return color
    }

    @objc public func colorizeView(_ image: UIImage?) {
        guard let view = view else {
            return
        }
        
        if colorMode == 1 {
            let color = UIColor(red: 1.0,
                 green: 0.0,
                 blue: 0.0,
                 alpha: dynamicColorAlpha)
            let scolor = UIColor(red: 0.0,
                 green: 1.0,
                 blue: 0.0,
                 alpha: dynamicColorAlpha)
            let sscolor = UIColor(red: 0.0,
                 green: 0.0,
                 blue: 1.0,
                 alpha: dynamicColorAlpha)
             
            view.updateWave(color,
                 subwaveColor: scolor,
                 subSubwaveColor: sscolor)
            
        }
        else if let image = image, colorMode == 0 {
            let color = getAverageColor(from: image, withAlpha: dynamicColorAlpha)
            calculatedColor = color
            view.updateWave(color, subwaveColor: color)
        }
        else {
            let color = waveColor!
            view.updateWave(color, subwaveColor: color)
        }
    }

    private func setDictionary(_ dict: Dictionary<String, Any>) {
        application = (dict["application"] as! String)
        enabled = dict["enabled"] as? Bool ?? true

        style = dict["style"] as? Int ?? 0
        colorMode = dict["colorMode"] as? Int ?? 0
        enableAutoUIColor = dict["enableAutoUIColor"] as? Bool ?? true
        ignoreColorFlow = dict["ignoreColorFlow"] as? Bool ?? false
        enableCoverArtBugFix = dict["enableCoverArtBugFix"] as? Bool ?? false
        disableBatterySaver = dict["disableBatterySaver"] as? Bool ?? false
        enableFFT = dict["enableFFT"] as? Bool ?? false
        enableAutoHide = dict["enableAutoHide"] as? Bool ?? true
        
        if let value = dict["waveColor"] as? UIColor {
            waveColor = value
        }
        else if let value = dict["waveColor"] as? String {
            waveColor = LCPParseColorString(value, "#000000:0.5")
        }
        else {
            waveColor = UIColor.black.withAlphaComponent(0.5)
        }
        
        if let value = dict["subwaveColor"] as? UIColor {
            subwaveColor = value
        }
        else if let value = dict["subwaveColor"] as? String {
            subwaveColor = LCPParseColorString(value, "#000000:0.5")
        }
        else {
            subwaveColor = UIColor.black.withAlphaComponent(0.5)
        }
        
        if let value = dict["subSubwaveColor"] as? UIColor {
            subwaveColor = value
        }
        else if let value = dict["subSubwaveColor"] as? String {
            subwaveColor = LCPParseColorString(value, "#000000:0.5")
        }
        else {
            subwaveColor = UIColor.black.withAlphaComponent(0.5)
        }
        
      gain = dict["gain"] as? Double ?? 50
      limiter = dict["limiter"] as? Double ?? 0
      numberOfPoints = dict["numberOfPoints"] as? Int ?? 8
      sensitivity = dict["sensitivity"] as? CGFloat ?? 1
      dynamicColorAlpha = dict["dynamicColorAlpha"] as? CGFloat ?? 0.6

      barSpacing = dict["barSpacing"] as? CGFloat ?? 5
      barCornerRadius = dict["barCornerRadius"] as? CGFloat ?? 0
      lineThickness = dict["lineThickness"] as? CGFloat ?? 5

      waveOffset = dict["waveOffset"] as? CGFloat ?? 0

      fps = dict["fps"] as? Int ?? 24

    }

    private class func parseConfig(forApplication name: String) -> Dictionary<String, Any> {
        var prefs: [String : Any] = [:]
        prefs["application"] = name

        let MSHFPrefsFile = "/var/mobile/Library/Preferences/com.ryannair05.mitsuhaforever.plist"

        if let file = NSDictionary(contentsOfFile: MSHFPrefsFile) as Dictionary? {
            for key in file.keys {
                guard let key = key as? String else {
                    continue
                }
                prefs[key] = file[key as NSObject]
            }
        }

        for key in prefs.keys {
            let removedKey = key.replacingOccurrences(of: "MSHF\(name)", with: "")
            let lowerCaseKey = "\(removedKey.prefix(1).lowercased())\(removedKey.dropFirst(1))"

            prefs[lowerCaseKey] = prefs[key]
        }

        prefs["subwaveColor"] = prefs["waveColor"]
        prefs["subSubwaveColor"] = prefs["waveColor"]

        return prefs
    }

    @objc private func reload() {
        let oldStyle = style
        setDictionary(MSHFConfig.parseConfig(forApplication: application ?? ""))
        guard let view = view else {
            return
        }
        if style != oldStyle {
            initializeView(withFrame: view.frame)
            view.start()
        } else {
            configureView()
        }
    }

    @objc public class func loadConfig(forApplication name: String) -> MSHFConfig {
        return MSHFConfig(
            dictionary: MSHFConfig.parseConfig(forApplication: name))
    }
}
